processor/metricstransformprocessor/metrics_transform_processor.go

```diff
@@ -0,0 +1,169 @@
+// Copyright 2020 OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package metricstransformprocessor
+
+import (
+	"context"
+	"log"
+
+	"go.opentelemetry.io/collector/component"
+	"go.opentelemetry.io/collector/consumer"
+	"go.opentelemetry.io/collector/consumer/pdata"
+	"go.opentelemetry.io/collector/consumer/pdatautil"
+
+	metricspb "github.com/census-instrumentation/opencensus-proto/gen-go/metrics/v1"
+)
+
+type metricsTransformProcessor struct {
+	cfg        *Config
+	next       consumer.MetricsConsumer
+	metricname string
+	action     ConfigAction
+	newname    string
+	operations []Operation
+}
+
+var _ component.MetricsProcessor = (*metricsTransformProcessor)(nil)
+
+func newMetricsTransformProcessor(next consumer.MetricsConsumer, cfg *Config) (*metricsTransformProcessor, error) {
+	return &metricsTransformProcessor{
+		cfg:        cfg,
+		next:       next,
+		metricname: cfg.MetricName,
+		action:     cfg.Action,
+		newname:    cfg.NewName,
+		operations: cfg.Operations,
+	}, nil
+}
+
+// GetCapabilities returns the Capabilities assocciated with the resource processor.
+func (mtp *metricsTransformProcessor) GetCapabilities() component.ProcessorCapabilities {
+	return component.ProcessorCapabilities{MutatesConsumedData: false}
+}
+
+// Start is invoked during service startup.
+func (*metricsTransformProcessor) Start(ctx context.Context, host component.Host) error {
+	return nil
+}
+
+// Shutdown is invoked during service shutdown.
+func (*metricsTransformProcessor) Shutdown(ctx context.Context) error {
+	return nil
+}
+
+// ConsumeMetrics implements the MetricsProcessor interface
+func (mtp *metricsTransformProcessor) ConsumeMetrics(ctx context.Context, md pdata.Metrics) error {
+	return mtp.next.ConsumeMetrics(ctx, mtp.transform(md))
+}
+
+// transform transforms the metrics based on the information specified in the config
+func (mtp *metricsTransformProcessor) transform(md pdata.Metrics) pdata.Metrics {
+	mds := pdatautil.MetricsToMetricsData(md)
+
+	for i, data := range mds {
+		// if the new name is not valid, discard this operation for this list of metrics
+		if mtp.validNewName(data.Metrics) {
+			for _, metric := range data.Metrics {
```

Comment:
Is there any performance risk with these for loops going through the metrics? Is there some way we could do all the operations with just a single pass through both the transforms and the metric names, or more cleverly match them together using a hash map?

Reply:
Since we have to look through the metrics at least once to locate the metrics that can be selected based on the name, I don't see how else we should look through them without looping through. The [implementation for the data new interface](https://github.com/open-telemetry/opentelemetry-collector/pull/1086/files) seems to also have similar implementation.

Reply:
Thanks. That is definitely a needed change if we are going to extend the config that way! I put that validation there because of a conversation I had with Dave about possible collision of metrics. Basically, we thought that it would be problematic if there are two metrics with the same name and same labels exist at the same time

Reply:
I think I assumed that the config would allow multiple metrics - I agree that it makes things much cleaner if you support that.

Reply:
Updated this, so that a list of transformations can be done with each on a separate metric. [Readme](https://github.com/JingboWangGoogle/opentelemetry-collector-contrib/tree/readme/processor/metricstransformprocessor) is also updated